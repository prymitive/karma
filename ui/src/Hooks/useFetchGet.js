import { useState, useEffect, useCallback, useRef } from "react";

import merge from "lodash.merge";

import promiseRetry from "promise-retry";

import { CommonOptions, FetchRetryConfig } from "Common/Fetch";

const useFetchGet = (
  uri,
  { autorun = true, deps = [], fetcher = null } = {}
) => {
  const [response, setResponse] = useState(null);
  const [error, setError] = useState(null);
  const [isLoading, setIsLoading] = useState(true);
  const [isRetrying, setIsRetrying] = useState(false);
  const [retryCount, setRetryCount] = useState(0);
  const isCanceled = useRef(false);

  const cancelGet = useCallback(() => {
    isCanceled.current = true;
  }, []);

  const get = useCallback(async () => {
    isCanceled.current = false;

    try {
      setIsLoading(true);
      setRetryCount(0);
      setError(null);
      const res = await promiseRetry(
        (retry, number) =>
          (fetcher || fetch)(
            uri,
            merge(
              {},
              {
                method: "GET",
              },
              CommonOptions,
              {
                mode: number <= FetchRetryConfig.retries ? "cors" : "no-cors",
              }
            )
          ).catch((err) => {
            if (!isCanceled.current) {
              setIsRetrying(true);
              setRetryCount(number);
              return retry(err);
            }
          }),
        FetchRetryConfig
      );

      if (!isCanceled.current) {
        let body;
        const contentType = res.headers.get("content-type");
        if (contentType && contentType.indexOf("application/json") !== -1) {
          body = await res.json();
        } else {
          body = await res.text();
        }

        if (!isCanceled.current) {
          if (res.ok) {
            setResponse(body);
          } else {
            setError(body);
          }
          setIsLoading(false);
          setIsRetrying(false);
        }
      }
    } catch (error) {
      setError(error.message);
      setIsLoading(false);
      setIsRetrying(false);
    }
  }, [uri, fetcher]);

  useEffect(() => {
    if (autorun) get();

    return () => cancelGet();
  }, [uri, get, cancelGet, autorun, ...deps]); // eslint-disable-line react-hooks/exhaustive-deps

  return { response, error, isLoading, isRetrying, retryCount, get, cancelGet };
};

export { useFetchGet };
