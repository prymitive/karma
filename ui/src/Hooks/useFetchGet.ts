import { useState, useEffect, useCallback, useRef } from "react";

import merge from "lodash.merge";

import promiseRetry from "promise-retry";

import { CommonOptions, FetchRetryConfig } from "Common/Fetch";

export type FetchFunctionT = (request: RequestInfo) => Promise<Response>;

export interface FetchGetOptionsT {
  autorun?: boolean;
  deps?: string[] | number[];
  fetcher?: null | FetchFunctionT;
}

interface ResponseState<T> {
  response: null | T;
  error: null | string;
  isLoading: boolean;
  isRetrying: boolean;
  retryCount: number;
}

export interface FetchGetResultT<T> {
  response: null | T;
  error: null | string;
  isLoading: boolean;
  isRetrying: boolean;
  retryCount: number;
  get: () => void;
  cancelGet: () => void;
}

const useFetchGet = <T>(
  uri: string,
  { autorun = true, deps = [], fetcher = null }: FetchGetOptionsT = {}
): FetchGetResultT<T> => {
  const [response, setResponse] = useState<ResponseState<T>>({
    response: null,
    error: null,
    isLoading: autorun,
    isRetrying: false,
    retryCount: 0,
  });
  const isCanceled = useRef<boolean>(false);

  const cancelGet = useCallback(() => {
    isCanceled.current = true;
  }, []);

  const get = useCallback(async () => {
    isCanceled.current = false;

    try {
      setResponse((r) => ({
        ...r,
        isLoading: true,
        isRetrying: false,
        retryCount: 0,
      }));

      const res = await promiseRetry(
        (retry: (err: Error) => Promise<Response>, n: number) =>
          (fetcher || fetch)(
            uri,
            merge(
              {},
              {
                method: "GET",
              },
              CommonOptions,
              {
                mode: n <= FetchRetryConfig.retries ? "cors" : "no-cors",
              }
            ) as RequestInit
          ).catch((err: Error) => {
            if (!isCanceled.current) {
              setResponse((r) => ({
                ...r,
                isRetrying: true,
                retryCount: n,
              }));
              return retry(err);
            }
          }),
        FetchRetryConfig
      );

      if (res !== undefined && !isCanceled.current) {
        let body;
        const contentType = res.headers.get("content-type");
        if (contentType && contentType.indexOf("application/json") !== -1) {
          body = await res.json();
        } else {
          body = await res.text();
        }

        if (!isCanceled.current) {
          if (res.ok) {
            setResponse({
              response: body,
              error: null,
              isLoading: false,
              isRetrying: false,
              retryCount: 0,
            });
          } else {
            setResponse({
              response: null,
              error: body,
              isLoading: false,
              isRetrying: false,
              retryCount: 0,
            });
          }
        }
      }
    } catch (error) {
      setResponse((r) => ({
        ...r,
        error: error.message,
        isLoading: false,
        isRetrying: false,
      }));
    }
  }, [uri, fetcher]);

  useEffect(() => {
    if (autorun) get();

    return () => cancelGet();
  }, [uri, get, cancelGet, autorun, ...deps]); // eslint-disable-line react-hooks/exhaustive-deps

  return { get, cancelGet, ...response };
};

export { useFetchGet };
