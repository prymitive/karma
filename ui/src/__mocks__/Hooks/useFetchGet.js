import { useState, useEffect, useCallback } from "react";

import { MockAPIResponse, MockSilenceResponse } from "__mocks__/Fetch";

const MockFetchStats = {
  getCalls: [],
  get calls() {
    return this.getCalls;
  },
  wasCalled(uri) {
    this.getCalls.push(uri);
  },
  reset() {
    this.getCalls = [];
    this.mockedData = {
      response: undefined,
      error: undefined,
      isLoading: undefined,
      isRetrying: undefined,
    };
  },
  mockedData: {
    response: undefined,
    error: undefined,
    isLoading: undefined,
    isRetrying: undefined,
  },
  setMockedData(data) {
    this.mockedData = data;
  },
};

const Mock = (uri, { autorun = true, deps = [] } = {}) => {
  const [response, setResponse] = useState(null);
  const [error] = useState(null);
  const [isLoading, setIsLoading] = useState(true);
  const [isRetrying] = useState(false);

  const cancelGet = useCallback(() => {}, []);

  const get = useCallback(() => {
    MockFetchStats.wasCalled(uri);

    const mockResponses = [
      // matcher name select suggestions
      {
        uri: "./labelNames.json",
        response: ["cluster", "job", "instance"],
      },
      // matcher value select suggestions
      {
        uri: "./labelValues.json?name=cluster",
        response: ["dev", "staging", "prod"],
      },
      // matcher value counters
      {
        re: /^\.\/alerts\.json\?q=/,
        response: MockAPIResponse(),
      },
      // silence browser
      {
        re: /^.\/silences\.json\?/,
        response: MockSilenceResponse("am", 14),
      },
      // filter autocomplete
      {
        uri: "./autocomplete.json?term=cluster",
        response: [
          "cluster=staging",
          "cluster=prod",
          "cluster=dev",
          "cluster!=staging",
          "cluster!=prod",
          "cluster!=dev",
        ],
      },
      {
        re: /^.\/autocomplete\.json\?term=/,
        response: ["foo=bar", "foo=~bar"],
      },
    ];

    for (const m of mockResponses) {
      if (m.re && uri.match(m.re)) {
        setResponse(m.response);
        setIsLoading(false);
        break;
      } else if (m.uri === uri) {
        setResponse(m.response);
        setIsLoading(false);
        break;
      }
    }
  }, [uri]);

  useEffect(() => {
    if (autorun) get();
    // eslint doesn't like ...deps
    // eslint-disable-next-line
  }, [uri, get, cancelGet, autorun, ...deps]);

  return {
    response:
      MockFetchStats.mockedData.response !== undefined
        ? MockFetchStats.mockedData.response
        : response,
    error:
      MockFetchStats.mockedData.error !== undefined
        ? MockFetchStats.mockedData.error
        : error,
    isLoading:
      MockFetchStats.mockedData.isLoading !== undefined
        ? MockFetchStats.mockedData.isLoading
        : isLoading,
    isRetrying:
      MockFetchStats.mockedData.isRetrying !== undefined
        ? MockFetchStats.mockedData.isRetrying
        : isRetrying,
    get,
    cancelGet,
  };
};

const useFetchGet = jest.fn(Mock);
useFetchGet.fetch = MockFetchStats;
useFetchGet.mockReset = () => {
  useFetchGet.mockClear();
  useFetchGet.mockImplementation(Mock);
  MockFetchStats.reset();
};

export { useFetchGet };
