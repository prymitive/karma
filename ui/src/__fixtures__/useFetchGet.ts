import { useState, useEffect, useCallback } from "react";

import { MockAPIResponse, MockSilenceResponse } from "__fixtures__/Fetch";
import { FetchGetResultT } from "Hooks/useFetchGet";
import { APIAlertsResponseT, APIManagedSilenceT } from "Models/APITypes";

type responseT = null | string[] | APIAlertsResponseT | APIManagedSilenceT[];

interface mockedDataT {
  response: undefined | responseT;
  error: undefined | null | string;
  isLoading: undefined | boolean;
  isRetrying: undefined | boolean;
  retryCount: undefined | number;
  get: () => void;
  cancelGet: () => void;
}

interface mockFetchStatsT {
  getCalls: string[];
  readonly calls: string[];
  wasCalled: (uri: string) => void;
  reset: () => void;
  mockedData: mockedDataT;
  setMockedData: (data: mockedDataT) => void;
}

const mockGet = jest.fn();
const mockCancelGet = jest.fn();

const MockFetchStats: mockFetchStatsT = {
  getCalls: [],
  get calls() {
    return this.getCalls;
  },
  wasCalled(uri: string) {
    this.getCalls.push(uri);
  },
  reset() {
    this.getCalls = [];
    this.mockedData = {
      response: undefined,
      error: undefined,
      isLoading: undefined,
      isRetrying: undefined,
      retryCount: undefined,
      get: mockGet,
      cancelGet: mockCancelGet,
    };
  },
  mockedData: {
    response: undefined,
    error: undefined,
    isLoading: undefined,
    isRetrying: undefined,
    retryCount: undefined,
    get: mockGet,
    cancelGet: mockCancelGet,
  },
  setMockedData(data: mockedDataT) {
    this.mockedData = data;
  },
};

const useFetchGetMock = (
  uri: string,
  { autorun = true, deps = [] } = {}
): FetchGetResultT<responseT> => {
  const [response, setResponse] = useState(null as responseT);
  const [error] = useState(null);
  const [isLoading, setIsLoading] = useState<boolean>(true);
  const [isRetrying] = useState<boolean>(false);

  const cancelGet = useCallback(() => {}, []);

  const get = useCallback(() => {
    MockFetchStats.wasCalled(uri);

    const mockResponses = [
      // matcher name select suggestions
      {
        uri: "./labelNames.json",
        response: ["cluster", "job", "instance"],
      },
      // matcher value select suggestions
      {
        uri: "./labelValues.json?name=cluster",
        response: ["dev", "staging", "prod"],
      },
      // matcher value counters
      {
        re: /^\.\/alerts\.json\?q=/,
        response: MockAPIResponse(),
      },
      // silence browser
      {
        re: /^.\/silences\.json\?/,
        response: MockSilenceResponse("am", 14),
      },
      // filter autocomplete
      {
        uri: "./autocomplete.json?term=cluster",
        response: [
          "cluster=staging",
          "cluster=prod",
          "cluster=dev",
          "cluster!=staging",
          "cluster!=prod",
          "cluster!=dev",
        ],
      },
      {
        re: /^.\/autocomplete\.json\?term=/,
        response: ["foo=bar", "foo=~bar"],
      },
    ];

    for (const m of mockResponses) {
      if (m.re && uri.match(m.re)) {
        setResponse(m.response);
        setIsLoading(false);
        break;
      } else if (m.uri === uri) {
        setResponse(m.response);
        setIsLoading(false);
        break;
      }
    }
  }, [uri]);

  useEffect(() => {
    if (autorun) get();
    // eslint doesn't like ...deps
    // eslint-disable-next-line
  }, [uri, get, cancelGet, autorun, ...deps]);

  return {
    response:
      MockFetchStats.mockedData.response !== undefined
        ? MockFetchStats.mockedData.response
        : response,
    error:
      MockFetchStats.mockedData.error !== undefined
        ? MockFetchStats.mockedData.error
        : error,
    isLoading:
      MockFetchStats.mockedData.isLoading !== undefined
        ? MockFetchStats.mockedData.isLoading
        : isLoading,
    isRetrying:
      MockFetchStats.mockedData.isRetrying !== undefined
        ? MockFetchStats.mockedData.isRetrying
        : isRetrying,
    retryCount:
      MockFetchStats.mockedData.retryCount !== undefined
        ? MockFetchStats.mockedData.retryCount
        : 0,
    get,
    cancelGet,
  };
};

useFetchGetMock.fetch = MockFetchStats;
useFetchGetMock._mockGet = mockGet;
useFetchGetMock._cancelGet = mockCancelGet;

export { useFetchGetMock };
