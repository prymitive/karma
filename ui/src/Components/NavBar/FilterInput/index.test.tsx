import { act } from "react-dom/test-utils";

import { mount, render } from "enzyme";

import toDiffableHtml from "diffable-html";

import { useFetchGetMock } from "__fixtures__/useFetchGet";
import { AlertStore, NewUnappliedFilter } from "Stores/AlertStore";
import { Settings } from "Stores/Settings";
import { FilterInput } from ".";

let alertStore: AlertStore;
let settingsStore: Settings;
let originalInnerWidth: number;

declare let global: any;

beforeAll(() => {
  originalInnerWidth = global.window.innerWidth;
});

beforeEach(() => {
  global.window.innerWidth = originalInnerWidth;
  alertStore = new AlertStore([]);
  settingsStore = new Settings(null);
  jest.useFakeTimers();
});

afterEach(() => {
  jest.restoreAllMocks();
  global.window.innerWidth = originalInnerWidth;
});

const MountedInput = () => {
  return mount(
    <FilterInput alertStore={alertStore} settingsStore={settingsStore} />
  );
};

describe("<FilterInput />", () => {
  it("matches snapshot with no filters", () => {
    const tree = render(
      <FilterInput alertStore={alertStore} settingsStore={settingsStore} />
    );
    expect(toDiffableHtml(tree.html())).toMatchSnapshot();
    expect(alertStore.filters.values).toHaveLength(0);
  });

  it("matches snapshot with some filters", () => {
    alertStore.filters.setFilterValues([
      NewUnappliedFilter("foo=bar"),
      NewUnappliedFilter("baz!=bar"),
    ]);
    const tree = render(
      <FilterInput alertStore={alertStore} settingsStore={settingsStore} />
    );
    expect(toDiffableHtml(tree.html())).toMatchSnapshot();
    expect(alertStore.filters.values).toHaveLength(2);
  });

  it("input gets focus by default on desktop", () => {
    global.window.innerWidth = 768;
    const tree = MountedInput();
    expect(
      tree.find("div.components-filterinput-outer").hasClass("bg-focused")
    ).toBe(true);
  });

  it("input doesn't get focus by default on mobile", () => {
    global.window.innerWidth = 767;
    const tree = MountedInput();
    expect(
      tree.find("div.components-filterinput-outer").hasClass("bg-focused")
    ).toBe(false);
  });

  it("onChange should modify inputStore.value", () => {
    const tree = MountedInput();
    tree.find("input").simulate("change", { target: { value: "foo=bar" } });
    act(() => {
      jest.runOnlyPendingTimers();
    });

    expect(
      tree.find("input.components-filterinput-wrapper").props().value
    ).toBe("foo=bar");
  });

  it("submit should modify alertStore.filters", () => {
    const tree = MountedInput();

    tree.find("input").simulate("change", { target: { value: "foo=bar" } });
    act(() => {
      jest.runOnlyPendingTimers();
    });

    expect(alertStore.filters.values).toHaveLength(0);
    tree.find("form").simulate("submit");
    expect(alertStore.filters.values).toHaveLength(1);
    expect(alertStore.filters.values[0]).toMatchObject(
      NewUnappliedFilter("foo=bar")
    );
  });

  it("submit should be no-op if input value is empty", () => {
    const tree = MountedInput();
    tree.find("input").simulate("change", { target: { value: "" } });
    act(() => {
      jest.runOnlyPendingTimers();
    });

    expect(alertStore.filters.values).toHaveLength(0);
    tree.find("form").simulate("submit");
    expect(alertStore.filters.values).toHaveLength(0);
  });

  it("clicking on form-control div focuses input", () => {
    const tree = MountedInput();
    const formControl = tree.find("div.form-control");
    formControl.simulate("click");
    //expect(tree.find("input:focus")).toHaveLength(1);
    expect(
      tree.find("div.components-filterinput-outer").hasClass("bg-focused")
    ).toBe(true);
  });

  it("clicking input changes background color", () => {
    const tree = MountedInput();
    tree.find("input").simulate("click");
    expect(toDiffableHtml(tree.html())).toMatch(/bg-focused/);
  });

  it("focusing input changes background color", () => {
    const tree = MountedInput();
    const input = tree.find("input");
    input.simulate("focus");
    expect(toDiffableHtml(tree.html())).toMatch(/bg-focused/);
  });

  it("focusing form changes background color", () => {
    const tree = MountedInput();
    const formControl = tree.find(".form-control");
    formControl.find("input").simulate("focus");
    expect(toDiffableHtml(tree.html())).toMatch(/bg-focused/);
  });

  it("bluring input changes background color", async () => {
    const tree = MountedInput();
    const formControl = tree.find(".form-control");
    formControl
      .find("input")
      .simulate("change", { target: { value: "cluster" } });
    formControl.find("input").simulate("blur");
    expect(toDiffableHtml(tree.html())).not.toMatch(/bg-focused/);
    tree.unmount();
  });
});

describe("<FilterInput autocomplete />", () => {
  it("fetches suggestions on input change", async () => {
    const tree = MountedInput();
    tree.find("input").simulate("change", { target: { value: "cluster" } });
    act(() => {
      jest.runOnlyPendingTimers();
    });

    expect(useFetchGetMock.fetch.calls).toHaveLength(1);
    expect(useFetchGetMock.fetch.calls[0]).toContain(
      "./autocomplete.json?term=cluster"
    );
    tree.unmount();
  });

  it("doesn't fetch any suggestion if the input value is empty", () => {
    const tree = MountedInput();
    tree.find("input").simulate("change", { target: { value: "" } });
    act(() => {
      jest.runOnlyPendingTimers();
    });
    expect(useFetchGetMock.fetch.calls).toHaveLength(0);
    tree.unmount();
  });

  it("highliting a suggestion makes it active", async () => {
    const tree = MountedInput();
    tree.find("input").simulate("change", { target: { value: "cluster" } });
    act(() => {
      jest.runOnlyPendingTimers();
    });

    // suggestions are rendered only when input is focused
    tree.find("input").simulate("focus");
    // find() doesn't pick up suggestions even when tree.html() shows them
    // forcing update seems to solve it
    // https://github.com/airbnb/enzyme/issues/1233#issuecomment-343449560
    tree.update();

    tree.find("input").simulate("keydown", { keyCode: 40, key: "ArrowDown" });
    tree.update();
    expect(tree.find(".dropdown-item").at(0).html()).toMatch(/active/);
  });

  it("clicking on a suggestion adds it to filters", async () => {
    const tree = MountedInput();
    tree.find("input").simulate("change", { target: { value: "cluster" } });
    act(() => {
      jest.runOnlyPendingTimers();
    });

    // suggestions are rendered only when input is focused
    tree.find("input").simulate("focus");

    // find() doesn't pick up suggestions even when tree.html() shows them
    // forcing update seems to solve it
    // https://github.com/airbnb/enzyme/issues/1233#issuecomment-343449560
    tree.update();
    const suggestion = tree.find(".dropdown-item").at(1);
    expect(suggestion.text()).toBe("cluster=prod");
    suggestion.simulate("click");
    expect(alertStore.filters.values).toHaveLength(1);
    expect(alertStore.filters.values[0]).toMatchObject({
      raw: "cluster=prod",
    });
  });

  it("handles failed suggestion fetches", async () => {
    useFetchGetMock.fetch.setMockedData({
      response: null,
      error: "fake error",
      isLoading: false,
      isRetrying: false,
      retryCount: 0,
      get: jest.fn(),
      cancelGet: jest.fn(),
    });

    const tree = MountedInput();
    tree.find("input").simulate("change", { target: { value: "cluster" } });
    act(() => {
      jest.runOnlyPendingTimers();
    });
  });
});
