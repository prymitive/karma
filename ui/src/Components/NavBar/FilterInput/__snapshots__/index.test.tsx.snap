// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FilterInput /> matches snapshot with no filters 1`] = `
"
<div class=\\"input-group w-100 me-2 components-filterinput-outer bg-transparent\\">
  <div class=\\"form-control components-filterinput border-0 rounded-0 bg-inherit\\">
    <div class=\\"autosuggest d-inline-block mw-100\\"
         role=\\"combobox\\"
         aria-haspopup=\\"listbox\\"
         aria-owns=\\"downshift-0-menu\\"
         aria-expanded=\\"false\\"
    >
      <span class=\\"input-group-text text-muted d-inline-block me-2 border-0 bg-inherit px-1\\">
        <svg aria-hidden=\\"true\\"
             focusable=\\"false\\"
             data-prefix=\\"fas\\"
             data-icon=\\"search\\"
             class=\\"svg-inline--fa fa-search fa-w-16 \\"
             role=\\"img\\"
             xmlns=\\"http://www.w3.org/2000/svg\\"
             viewbox=\\"0 0 512 512\\"
        >
          <path fill=\\"currentColor\\"
                d=\\"M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z\\"
          >
          </path>
        </svg>
      </span>
      <input class=\\"components-filterinput-wrapper text-white mw-100\\"
             size=\\"1\\"
             id=\\"downshift-0-input\\"
             aria-autocomplete=\\"list\\"
             aria-controls=\\"downshift-0-menu\\"
             aria-labelledby=\\"downshift-0-label\\"
             autocomplete=\\"off\\"
             value
      >
    </div>
    <span class=\\"dropdown\\"
          id=\\"downshift-0-menu\\"
          role=\\"listbox\\"
          aria-labelledby=\\"downshift-0-label\\"
    >
    </span>
  </div>
  <span class=\\"input-group-text border-0 rounded-0 bg-inherit px-0\\">
    <button class=\\"btn border-0 rounded-0 bg-inherit cursor-pointer components-navbar-history px-2 py-0 components-navbar-icon\\"
            type=\\"button\\"
            data-toggle=\\"dropdown\\"
            aria-haspopup=\\"true\\"
            aria-expanded=\\"true\\"
    >
      <svg aria-hidden=\\"true\\"
           focusable=\\"false\\"
           data-prefix=\\"fas\\"
           data-icon=\\"caret-down\\"
           class=\\"svg-inline--fa fa-caret-down fa-w-10 \\"
           role=\\"img\\"
           xmlns=\\"http://www.w3.org/2000/svg\\"
           viewbox=\\"0 0 320 512\\"
      >
        <path fill=\\"currentColor\\"
              d=\\"M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z\\"
        >
        </path>
      </svg>
    </button>
  </span>
</div>
"
`;

exports[`<FilterInput /> matches snapshot with some filters 1`] = `
"
<div class=\\"input-group w-100 me-2 components-filterinput-outer bg-transparent\\">
  <div class=\\"form-control components-filterinput border-0 rounded-0 bg-inherit\\">
    <button type=\\"button\\"
            class=\\"btn-secondary btn-sm components-filteredinputlabel components-label btn d-inline-flex flex-row align-items-center\\"
    >
      <svg aria-hidden=\\"true\\"
           focusable=\\"false\\"
           data-prefix=\\"fas\\"
           data-icon=\\"spinner\\"
           class=\\"svg-inline--fa fa-spinner fa-w-16 fa-spin \\"
           role=\\"img\\"
           xmlns=\\"http://www.w3.org/2000/svg\\"
           viewbox=\\"0 0 512 512\\"
      >
        <path fill=\\"currentColor\\"
              d=\\"M304 48c0 26.51-21.49 48-48 48s-48-21.49-48-48 21.49-48 48-48 48 21.49 48 48zm-48 368c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zm208-208c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zM96 256c0-26.51-21.49-48-48-48S0 229.49 0 256s21.49 48 48 48 48-21.49 48-48zm12.922 99.078c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.491-48-48-48zm294.156 0c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.49-48-48-48zM108.922 60.922c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.491-48-48-48z\\"
        >
        </path>
      </svg>
      <div class=\\"components-filteredinputlabel-text flex-grow-1 flex-shrink-1 ms-1 tooltip-trigger\\">
        <span tabindex=\\"0\\"
              class=\\"cursor-text px-1\\"
        >
          foo=bar
        </span>
      </div>
      <svg aria-hidden=\\"true\\"
           focusable=\\"false\\"
           data-prefix=\\"fas\\"
           data-icon=\\"times\\"
           class=\\"svg-inline--fa fa-times fa-w-11 cursor-pointer text-reset ms-1\\"
           role=\\"img\\"
           xmlns=\\"http://www.w3.org/2000/svg\\"
           viewbox=\\"0 0 352 512\\"
      >
        <path fill=\\"currentColor\\"
              d=\\"M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z\\"
        >
        </path>
      </svg>
    </button>
    <button type=\\"button\\"
            class=\\"btn-secondary btn-sm components-filteredinputlabel components-label btn d-inline-flex flex-row align-items-center\\"
    >
      <svg aria-hidden=\\"true\\"
           focusable=\\"false\\"
           data-prefix=\\"fas\\"
           data-icon=\\"spinner\\"
           class=\\"svg-inline--fa fa-spinner fa-w-16 fa-spin \\"
           role=\\"img\\"
           xmlns=\\"http://www.w3.org/2000/svg\\"
           viewbox=\\"0 0 512 512\\"
      >
        <path fill=\\"currentColor\\"
              d=\\"M304 48c0 26.51-21.49 48-48 48s-48-21.49-48-48 21.49-48 48-48 48 21.49 48 48zm-48 368c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zm208-208c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zM96 256c0-26.51-21.49-48-48-48S0 229.49 0 256s21.49 48 48 48 48-21.49 48-48zm12.922 99.078c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.491-48-48-48zm294.156 0c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.49-48-48-48zM108.922 60.922c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.491-48-48-48z\\"
        >
        </path>
      </svg>
      <div class=\\"components-filteredinputlabel-text flex-grow-1 flex-shrink-1 ms-1 tooltip-trigger\\">
        <span tabindex=\\"0\\"
              class=\\"cursor-text px-1\\"
        >
          baz!=bar
        </span>
      </div>
      <svg aria-hidden=\\"true\\"
           focusable=\\"false\\"
           data-prefix=\\"fas\\"
           data-icon=\\"times\\"
           class=\\"svg-inline--fa fa-times fa-w-11 cursor-pointer text-reset ms-1\\"
           role=\\"img\\"
           xmlns=\\"http://www.w3.org/2000/svg\\"
           viewbox=\\"0 0 352 512\\"
      >
        <path fill=\\"currentColor\\"
              d=\\"M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z\\"
        >
        </path>
      </svg>
    </button>
    <div class=\\"autosuggest d-inline-block mw-100\\"
         role=\\"combobox\\"
         aria-haspopup=\\"listbox\\"
         aria-owns=\\"downshift-1-menu\\"
         aria-expanded=\\"false\\"
    >
      <input class=\\"components-filterinput-wrapper text-white mw-100\\"
             size=\\"1\\"
             id=\\"downshift-1-input\\"
             aria-autocomplete=\\"list\\"
             aria-controls=\\"downshift-1-menu\\"
             aria-labelledby=\\"downshift-1-label\\"
             autocomplete=\\"off\\"
             value
      >
    </div>
    <span class=\\"dropdown\\"
          id=\\"downshift-1-menu\\"
          role=\\"listbox\\"
          aria-labelledby=\\"downshift-1-label\\"
    >
    </span>
  </div>
  <span class=\\"input-group-text border-0 rounded-0 bg-inherit px-0\\">
    <button class=\\"btn border-0 rounded-0 bg-inherit cursor-pointer components-navbar-history px-2 py-0 components-navbar-icon\\"
            type=\\"button\\"
            data-toggle=\\"dropdown\\"
            aria-haspopup=\\"true\\"
            aria-expanded=\\"true\\"
    >
      <svg aria-hidden=\\"true\\"
           focusable=\\"false\\"
           data-prefix=\\"fas\\"
           data-icon=\\"caret-down\\"
           class=\\"svg-inline--fa fa-caret-down fa-w-10 \\"
           role=\\"img\\"
           xmlns=\\"http://www.w3.org/2000/svg\\"
           viewbox=\\"0 0 320 512\\"
      >
        <path fill=\\"currentColor\\"
              d=\\"M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z\\"
        >
        </path>
      </svg>
    </button>
  </span>
</div>
"
`;
